internal/controller/templatechain_controller.go

```diff
@@ -145,70 +169,144 @@ func (r *TemplateChainReconciler) ReconcileTemplateChain(ctx context.Context, te
 	switch templateKind {
 	case hmc.ClusterTemplateKind:
 		ctList := &hmc.ClusterTemplateList{}
-		err := cl.List(ctx, ctList)
+		err := r.Client.List(ctx, ctList, opts)
 		if err != nil {
-			return nil, nil, err
+			return nil, err
 		}
 		for _, template := range ctList.Items {
 			templates = append(templates, &template)
 		}
 	case hmc.ServiceTemplateKind:
 		stList := &hmc.ServiceTemplateList{}
-		err := cl.List(ctx, stList)
+		err := r.Client.List(ctx, stList, opts)
 		if err != nil {
-			return nil, nil, err
+			return nil, err
 		}
 		for _, template := range stList.Items {
 			templates = append(templates, &template)
 		}
 	default:
-		return nil, nil, fmt.Errorf("invalid Template kind. Supported kinds are %s and %s", hmc.ClusterTemplateKind, hmc.ServiceTemplateKind)
+		return nil, fmt.Errorf("invalid Template kind. Supported kinds are %s and %s", hmc.ClusterTemplateKind, hmc.ServiceTemplateKind)
 	}
+	return templates, nil
+}
 
-	systemTemplates = make(map[string]templateCommon, len(templates))
-	managedTemplates = make([]templateCommon, 0, len(templates))
-	for _, template := range templates {
-		if template.GetNamespace() == systemNamespace {
-			systemTemplates[template.GetName()] = template
+func (r TemplateChainReconciler) removeOrphanedTemplates(ctx context.Context, chain templateChain) error {
+	l := log.FromContext(ctx)
+
+	managedTemplates, err := r.getTemplates(ctx, chain.TemplateKind(), &client.ListOptions{
+		Namespace:     chain.GetNamespace(),
+		LabelSelector: labels.SelectorFromSet(map[string]string{hmc.HMCManagedLabelKey: hmc.HMCManagedLabelValue}),
+	})
+	if err != nil {
+		return err
+	}
+
+	// Removing templates not managed by any chain
+	var errs error
+	for _, template := range managedTemplates {
+		isOrphaned, err := isTemplateOrphaned(ctx, r.Client, chain.TemplateKind(), template.GetNamespace(), template.GetName())
+		if err != nil {
+			errs = errors.Join(errs, err)
 			continue
 		}
-
-		labels := template.GetLabels()
-		if template.GetNamespace() == targetNamespace &&
-			labels[hmc.HMCManagedLabelKey] == hmc.HMCManagedLabelValue &&
-			labels[HMCManagedByChainLabelKey] == templateChainName {
-			managedTemplates = append(managedTemplates, template)
+		if isOrphaned {
+			ll := l.WithValues("template kind", chain.TemplateKind(), "template namespace", template.GetNamespace(), "template name", template.GetName())
+			ll.Info("Deleting Template")
+
+			if err := r.Client.Delete(ctx, template); client.IgnoreNotFound(err) != nil {
+				errs = errors.Join(errs, err)
+				continue
+			}
+			ll.Info("Template has been deleted")
 		}
 	}
+	if errs != nil {
+		return errs
+	}
+	return nil
+}
 
-	return systemTemplates, managedTemplates, nil
+func isTemplateOrphaned(ctx context.Context, cl client.Client, templateKind, namespace, templateName string) (bool, error) {
+	opts := &client.ListOptions{
+		Namespace:     namespace,
+		FieldSelector: fields.SelectorFromSet(fields.Set{hmc.SupportedTemplateKey: templateName}),
+	}
```

Comment:
Yes, it would be great if I could list only non-deleted objects, limited to a single entry. But as far as I know, there’s no option to do that (I mean, to list items without a deletion timestamp only). Do you know of any? Using a custom indexer for the deletion timestamp doesn’t seem like good practice.

Reply:
Right, I missed that one, thanks for the clarification. Closing the conv